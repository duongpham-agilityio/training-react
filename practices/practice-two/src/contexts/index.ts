export { default as ToastProvider } from 'contexts/Toast';
export { default as PopupProvider } from 'contexts/Popup';
