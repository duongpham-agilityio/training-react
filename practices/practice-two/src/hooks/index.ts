export * from './useBookForm';
export * from './useSearchParam';
export * from './usePagination';
export * from './useFilter';
export * from './useDebounce';
export * from './useBooks';
export * from './useBook';
export * from './usePopupContext';
export * from './useToastContext';
