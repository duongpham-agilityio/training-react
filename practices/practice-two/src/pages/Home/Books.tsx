import { MouseEvent, ReactNode, memo, useCallback } from 'react';
import isEqual from 'react-fast-compare';

// Components
import { Heading } from 'components/commons';
import { Card } from 'components';

// Constants
import { ENDPOINT, MESSAGES } from '@constants';

// Helpers
import { convertDateTimeToTimeString } from 'helpers';

// Styles
import homeStyles from 'pages/Home/index.module.css';

// Types
import { Book } from 'types';

export interface BooksProps {
  books: Book[];
  onDeleteBook: (id: number) => void;
  children?: ReactNode;
}

const Books = ({ books, children, onDeleteBook }: BooksProps) => {
  const renderBook = useCallback(
    (book: Book) => {
      const { id, name, description, createdAt, imageURL } = book;
      const deleteHandler = (event: MouseEvent) => {
        event.preventDefault();

        onDeleteBook(id || 0);
      };

      return (
        <Card
          href={`/${ENDPOINT.BOOKS}/${id}`}
          title={name}
          description={description}
          publishedDate={convertDateTimeToTimeString(createdAt)}
          imageUrl={imageURL}
          key={id}
          onDeleteBook={deleteHandler}
        />
      );
    },
    [onDeleteBook]
  );

  return (
    <div className={homeStyles.content}>
      {books.length ? (
        <>
          <div className={homeStyles.grid}>{books.map(renderBook)}</div>
          {children}
        </>
      ) : (
        <div className={homeStyles.empty}>
          <Heading
            label={MESSAGES.EMPTY_TITLE}
            className={homeStyles.emptyTitle}
            size="xl"
          />
          <p className={homeStyles.emptyDescription}>
            {MESSAGES.EMPTY_DESCRIPTION}
          </p>
        </div>
      )}
    </div>
  );
};

const areCompare = (prev: BooksProps, next: BooksProps): boolean =>
  isEqual(prev.books, next.books) && isEqual(prev.children, next.children);

export default memo(Books, areCompare);
