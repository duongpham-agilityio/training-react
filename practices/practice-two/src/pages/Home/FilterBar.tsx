import { MouseEvent, ChangeEvent, memo, useCallback } from 'react';
import { Link } from 'react-router-dom';
import isEqual from 'react-fast-compare';

// Component
import { Button, Input } from 'components/commons';

// Styles
import homeStyles from 'pages/Home/index.module.css';

// Assets
import SearchIcon from 'assets/icons/search.svg';
import AddIcon from 'assets/icons/add.svg';

export interface SortOption {
  href: string;
  isActive: boolean;
  title: string;
}

export interface FilterBarProps {
  searchValue: string;
  sortOptions: SortOption[];
  onChangeSearch: (value: string) => void;
  onAddBook: (e: MouseEvent) => void;
}

const FilterBar = ({
  searchValue,
  sortOptions,
  onChangeSearch,
  onAddBook,
}: FilterBarProps) => {
  const changeSearchValue = useCallback(
    (event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
      onChangeSearch(event.target.value);
    },
    [onChangeSearch]
  );

  const renderOption = useCallback((option: SortOption) => {
    const { href, title, isActive } = option;

    return (
      <li key={title}>
        <Link
          to={href}
          className={`${homeStyles.navLink} ${
            isActive ? homeStyles.active : ''
          }`}
        >
          {title}
        </Link>
      </li>
    );
  }, []);

  return (
    <div className={homeStyles.navbar}>
      <Input
        className={homeStyles.search}
        value={searchValue}
        leftIcon={SearchIcon}
        placeholder="Search something..."
        onChange={changeSearchValue}
      />
      <ul className={homeStyles.navList}>{sortOptions.map(renderOption)}</ul>
      <Button
        label="Create"
        variant="primary"
        leftIcon={AddIcon}
        width="w-lg"
        border="b-lg"
        onClick={onAddBook}
      />
    </div>
  );
};

const areCompare = (prev: FilterBarProps, next: FilterBarProps): boolean =>
  isEqual(prev.sortOptions, next.sortOptions) &&
  isEqual(prev.searchValue, next.searchValue);

export default memo(FilterBar, areCompare);
