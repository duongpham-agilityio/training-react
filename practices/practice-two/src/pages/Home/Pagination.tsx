import { memo } from 'react';
import isEqual from 'react-fast-compare';

// Components
import { Button } from 'components/commons';

// Styles
import homeStyles from 'pages/Home/index.module.css';

export interface PaginationProps {
  currentPage: number;
  pagination: number[];
  onChangePage: (page: number) => void;
}

const Pagination = ({
  pagination,
  currentPage,
  onChangePage,
}: PaginationProps) => {
  return (
    <div className={homeStyles.pagination}>
      {pagination.map((_, index) => {
        const page = index + 1;
        const handleChangePage = () => {
          onChangePage(page);
        };

        return (
          <Button
            label={`${page}`}
            variant={currentPage === page ? 'primary' : 'secondary'}
            size="small"
            key={index}
            onClick={handleChangePage}
            data-testid="pagination"
          />
        );
      })}
    </div>
  );
};

const areCompare = (prev: PaginationProps, next: PaginationProps): boolean =>
  isEqual(prev.currentPage, next.currentPage) &&
  isEqual(prev.pagination, next.pagination);

export default memo(Pagination, areCompare);
