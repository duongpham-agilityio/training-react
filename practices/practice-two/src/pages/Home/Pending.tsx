import { memo } from 'react';

// Components
import { CardSkeleton } from 'components';

// Styles
import homeStyles from 'pages/Home/index.module.css';

const Pending = () => (
  <div className={homeStyles.content}>
    <div className={homeStyles.grid}>
      {Array.from({ length: 6 }).map((_, index) => (
        <CardSkeleton key={index} />
      ))}
    </div>
  </div>
);

export default memo(Pending);
