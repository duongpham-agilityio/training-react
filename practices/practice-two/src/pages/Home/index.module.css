.home {
  font-family: var(--primary-font);
}

.navbar {
  display: flex;
  gap: 15px;
  align-items: center;
  padding: 20px 0 0;
}

.search {
  display: none;
  width: 100%;
}

.navList {
  flex: 1;
  display: flex;
  gap: 17px;
}

.navLink {
  text-decoration: none;
  font-size: 14px;
  font-weight: 700;
  color: var(--primary-color);
}

.navLink.active {
  color: var(--secondary-color);
}

.content {
  padding: 27px 0;
}

.grid {
  display: grid;
  grid-template-columns: 1fr;
  gap: 22px;
}

.pagination {
  display: flex;
  gap: 16px;
  align-items: center;
  justify-content: center;
  padding: 40px 0;
}

.empty {
  height: 50vh;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.emptyDescription {
  font-size: 30px;
  color: var(--secondary-text-color);
}

/* Small */
@media (min-width: 576px) {
  .navLink {
    font-size: 18px;
  }
}

/* Medium */
@media (min-width: 768px) {
  .navList {
    gap: 27px;
  }

  .grid {
    grid-template-columns: 1fr 1fr;
  }

  .search {
    display: flex;
    width: 250px;
  }
}

/* Large */
@media (min-width: 992px) {
}

/* Extra-large */
@media (min-width: 1200px) {
  .grid {
    grid-template-columns: 1fr 1fr 1fr;
  }

  .search {
    width: 340px;
  }

  .navLink {
    font-size: 22px;
  }
}

/* Extra-extra-large */
@media (min-width: 1400px) {
}
