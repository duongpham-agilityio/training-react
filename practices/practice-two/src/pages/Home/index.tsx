import { memo, useCallback, useState } from 'react';

// Hooks
import { useBooks, useDebounce } from 'hooks';

// HOCs
import { withErrorBoundaries } from 'hocs/withErrorBoundaries';

//Components
import { Error } from 'components';
import Books from './Books';
import Pagination from './Pagination';
import Pending from './Pending';
import FilterBar, { SortOption } from './FilterBar';

// Constants
import { ENDPOINT, SORT, TITLE } from '@constants';

// Styles
import homeStyles from 'pages/Home/index.module.css';
import commonStyles from 'styles/commons/index.module.css';

// HOCs
import { WithUseFormProps, withUseForm } from 'hocs';

const Home = ({ dispatchAction }: WithUseFormProps) => {
  const {
    isLoading,
    param: { sort, page },
    error,
    pagination,
    data,
    setSearchParam,
    convertSearchParamsToString,
    deleteBook,
  } = useBooks();
  const [search, setSearch] = useState('');
  const debounce = useDebounce((value: string) =>
    setSearchParam('name', value)
  );

  const sortOptions: SortOption[] = [
    {
      href: `/${ENDPOINT.BOOKS}${convertSearchParamsToString(
        'sort',
        sort === SORT.ASCENDING ? '' : SORT.ASCENDING
      )}`,
      isActive: sort === SORT.ASCENDING,
      title: 'Ascending',
    },
    {
      href: `/${ENDPOINT.BOOKS}${convertSearchParamsToString(
        'sort',
        sort === SORT.DESCENDING ? '' : SORT.DESCENDING
      )}`,
      isActive: sort === SORT.DESCENDING,
      title: 'Descending',
    },
  ];

  const changeSearchData = useCallback(
    (value: string) => {
      setSearch(value);
      debounce(value);
    },
    [debounce]
  );

  const createBookHandler = useCallback(() => {
    dispatchAction({
      formData: {
        author: '',
        description: '',
        imageURL: '',
        name: '',
        createdAt: new Date().getTime(),
        deletedAt: null,
        updatedAt: new Date().getTime(),
      },
      title: TITLE.FORM_CREATE,
      type: 'create',
    });
  }, [dispatchAction]);

  const changePage = useCallback(
    (page: number) => {
      setSearchParam('page', `${page}`);
    },
    [setSearchParam]
  );

  if (error) {
    return <Error />;
  }

  return (
    <main className={commonStyles.container}>
      <section className={homeStyles.home}>
        <FilterBar
          searchValue={search}
          sortOptions={sortOptions}
          onChangeSearch={changeSearchData}
          onAddBook={createBookHandler}
        />

        {isLoading ? (
          <Pending />
        ) : (
          <Books books={data} onDeleteBook={deleteBook}>
            <Pagination
              onChangePage={changePage}
              currentPage={page}
              pagination={pagination}
            />
          </Books>
        )}
      </section>
    </main>
  );
};

export default memo(withErrorBoundaries(withUseForm(Home)));
