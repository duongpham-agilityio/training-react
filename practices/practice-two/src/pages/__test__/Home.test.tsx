// Mocks
import 'components/__test__/mocks/image.test';

import { act, fireEvent, render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { ReactNode } from 'react';

// Hooks
import * as hooks from 'hooks';

// Components
import Home from 'pages/Home';

// Providers
import { ToastProvider } from 'contexts';

// Mocks data
import { books } from 'mock-data';

// constants
import { MESSAGES, SORT } from '@constants';

const Wrapper = ({ children }: { children: ReactNode }) => (
  <BrowserRouter>
    <ToastProvider>{children}</ToastProvider>
  </BrowserRouter>
);

const setup = () =>
  render(<Home />, {
    wrapper: Wrapper,
  });

jest.mock('hooks', () => ({
  ...jest.requireActual('hooks'),
}));

const mockData: ReturnType<typeof hooks.useBooks> = {
  param: { sort: '', page: 1, name: 'html' },
  deleteBook: jest.fn(),
  data: books,
  pagination: [0, 1, 2, 3],
  setSearchParam: jest.fn(),
  convertSearchParamsToString: jest.fn(),
  error: undefined,
  mutate: jest.fn(),
  isValidating: false,
  isLoading: false,
};

describe('Home page', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  it('Match to snapshot', () => {
    jest.spyOn(hooks, 'useBooks').mockReturnValue(mockData);

    const { container } = setup();

    expect(container).toMatchSnapshot();
  });

  it('Error fetching ', () => {
    jest.spyOn(hooks, 'useBooks').mockReturnValue({ ...mockData, error: true });

    const { getByText } = setup();

    expect(getByText(MESSAGES.ERROR_TITLE)).toBeInTheDocument();
  });

  it('Load data', () => {
    jest
      .spyOn(hooks, 'useBooks')
      .mockReturnValue({ ...mockData, isLoading: true });
    const { getAllByTestId } = setup();

    expect(getAllByTestId('card-skeleton').length).toBeGreaterThan(0);
  });

  it('Render data', () => {
    jest.spyOn(hooks, 'useBooks').mockReturnValue({
      ...mockData,
      param: { sort: SORT.DESCENDING, page: 1, name: 'html' },
    });

    const { getByPlaceholderText } = setup();
    const input = getByPlaceholderText('Search something...');

    act(() => {
      fireEvent.change(input, {
        target: {
          value: 'html',
        },
      });
    });

    expect((input as HTMLInputElement).value).toBe('html');
  });

  it('Create book', async () => {
    jest.spyOn(hooks, 'useBooks').mockReturnValue({
      ...mockData,
      param: { sort: SORT.ASCENDING, page: 1, name: 'html' },
    });

    const { getAllByRole } = setup();
    const buttons = getAllByRole('button');

    act(() => {
      fireEvent.click(buttons[2]);
    });
  });

  it('Change page', () => {
    jest.spyOn(hooks, 'useBooks').mockReturnValue(mockData);

    const { getAllByTestId } = setup();
    const buttons = getAllByTestId('pagination');

    act(() => {
      fireEvent.click(buttons[1]);
    });
  });
});
