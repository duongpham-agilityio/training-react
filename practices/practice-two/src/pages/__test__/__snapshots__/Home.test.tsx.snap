// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page Match to snapshot 1`] = `
<div>
  <main>
    <section>
      <div>
        <div
          class="undefined function search() { [native code] }"
        >
          <button
            aria-label="Search button"
            class="undefined undefined
  undefined undefined
  undefined "
            style="background-color: transparent; padding: 0px;"
            type="button"
          >
            <img
              alt="This is the left icon for the button"
              height="25px"
              src="assets/icons/search.svg"
              width="22px"
            />
          </button>
          <input
            placeholder="Search something..."
            type="text"
            value=""
          />
          <button
            aria-label="The right button"
            class="undefined undefined
  undefined undefined
  undefined "
            style="background-color: transparent; padding: 0px;"
            type="button"
          />
        </div>
        <ul>
          <li>
            <a
              class="undefined "
              href="/booksundefined"
            >
              Ascending
            </a>
          </li>
          <li>
            <a
              class="undefined "
              href="/booksundefined"
            >
              Descending
            </a>
          </li>
        </ul>
        <button
          class="undefined undefined
  undefined undefined
  undefined "
          type="button"
        >
          <img
            alt="This is the left icon for the button"
            height="25px"
            src="assets/icons/add.svg"
            width="22px"
          />
          Create
        </button>
      </div>
      <div>
        <div>
          <a
            href="/books/1"
          >
            <div>
              <h2
                class="undefined undefined "
              >
                HTML/CSS Ebook 1
              </h2>
              <p>
                Description of some book will displayed here
              </p>
              <div>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/next.svg"
                    width="22px"
                  />
                </button>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/trash.svg"
                    width="22px"
                  />
                </button>
              </div>
            </div>
            <div>
              <p>
                1:39 PM
              </p>
              <img
                alt="HTML/CSS Ebook 1"
                src="http://books.google.com/books/content?id=KzzXzqLzXi8C&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api"
              />
            </div>
          </a>
          <a
            href="/books/3"
          >
            <div>
              <h2
                class="undefined undefined "
              >
                HTML/CSS Ebook dd 
              </h2>
              <p>
                Description of some book will displayed hereDescription of some book will displayed hereDescription of some book will displayed hereDescription of some book will displayed heresdssdsdsdsdDescription of some book will displayed hereDescription of some book will displayed hereDescription of some book will displayed hereDescription of some book will displayed here
              </p>
              <div>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/next.svg"
                    width="22px"
                  />
                </button>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/trash.svg"
                    width="22px"
                  />
                </button>
              </div>
            </div>
            <div>
              <p>
                1:39 PM
              </p>
              <img
                alt="HTML/CSS Ebook dd "
                src="http://books.google.com/books/content?id=KzzXzqLzXi8C&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api"
              />
            </div>
          </a>
          <a
            href="/books/4"
          >
            <div>
              <h2
                class="undefined undefined "
              >
                HTML/CSS Ebook 4
              </h2>
              <p>
                Description of some book will displayed here ds s ssss   sss    
              </p>
              <div>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/next.svg"
                    width="22px"
                  />
                </button>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/trash.svg"
                    width="22px"
                  />
                </button>
              </div>
            </div>
            <div>
              <p>
                1:39 PM
              </p>
              <img
                alt="HTML/CSS Ebook 4"
                src="http://books.google.com/books/content?id=KzzXzqLzXi8C&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api"
              />
            </div>
          </a>
          <a
            href="/books/7"
          >
            <div>
              <h2
                class="undefined undefined "
              >
                HTML/CSS Ebook 6
              </h2>
              <p>
                Description of some book will displayed here
              </p>
              <div>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/next.svg"
                    width="22px"
                  />
                </button>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/trash.svg"
                    width="22px"
                  />
                </button>
              </div>
            </div>
            <div>
              <p>
                1:39 PM
              </p>
              <img
                alt="HTML/CSS Ebook 6"
                src="http://books.google.com/books/content?id=KzzXzqLzXi8C&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api"
              />
            </div>
          </a>
          <a
            href="/books/10"
          >
            <div>
              <h2
                class="undefined undefined "
              >
                LIFE
              </h2>
              <p>
                LIFE Magazine is the treasured photographic magazine that chronicled the 20th Century. It now lives on at LIFE.com, the largest, most amazing collection of professional photography on the internet. Users can browse, search and view photos of todayâ€™s people and events. They have free access to share, print and post images for personal use.
              </p>
              <div>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/next.svg"
                    width="22px"
                  />
                </button>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/trash.svg"
                    width="22px"
                  />
                </button>
              </div>
            </div>
            <div>
              <p>
                5:22 PM
              </p>
              <img
                alt="LIFE"
                src="http://books.google.com/books/content?id=Mz8EAAAAMBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
              />
            </div>
          </a>
          <a
            href="/books/12"
          >
            <div>
              <h2
                class="undefined undefined "
              >
                The Rotarian
              </h2>
              <p>
                Established in 1911, The Rotarian is the official magazine of Rotary International and is circulated worldwide. Each issue contains feature articles, columns, and departments about, or of interest to, Rotarians. Seventeen Nobel Prize winners and 19 Pulitzer Prize winners â€“ from Mahatma Ghandi to Kurt Vonnegut Jr. â€“ have written for the magazine.
              </p>
              <div>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/next.svg"
                    width="22px"
                  />
                </button>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/trash.svg"
                    width="22px"
                  />
                </button>
              </div>
            </div>
            <div>
              <p>
                11:5 PM
              </p>
              <img
                alt="The Rotarian"
                src="http://books.google.com/books/content?id=ckEEAAAAMBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
              />
            </div>
          </a>
          <a
            href="/books/13"
          >
            <div>
              <h2
                class="undefined undefined "
              >
                Hello
              </h2>
              <p>
                asdasdasdasd
              </p>
              <div>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/next.svg"
                    width="22px"
                  />
                </button>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/trash.svg"
                    width="22px"
                  />
                </button>
              </div>
            </div>
            <div>
              <p>
                8:38 AM
              </p>
              <img
                alt="Hello"
                src="http://books.google.com/books/content?id=WNSUweeeSIIC&printsec=frontcover&img=1&zoom=1&source=gbs_api"
              />
            </div>
          </a>
          <a
            href="/books/14"
          >
            <div>
              <h2
                class="undefined undefined "
              >
                Popular Mechanics2333
              </h2>
              <p>
                Popular Mechanics inspires, instructs and influences readers to help them master the modern world. Whether itâ€™s practical DIY home-improvement tips, gadgets and digital technology, information on the newest cars or the latest breakthroughs in science -- PM is the ultimate guide to our high-tech lifestyle.
              </p>
              <div>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/next.svg"
                    width="22px"
                  />
                </button>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/trash.svg"
                    width="22px"
                  />
                </button>
              </div>
            </div>
            <div>
              <p>
                8:39 AM
              </p>
              <img
                alt="Popular Mechanics2333"
                src="http://books.google.com/books/content?id=QtQDAAAAMBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
              />
            </div>
          </a>
          <a
            href="/books/15"
          >
            <div>
              <h2
                class="undefined undefined "
              >
                Hello! Flex 4
              </h2>
              <p>
                Flex 4 is an open-source tool that allows developers to easily add life to web applications with dynamic user features, colorful transitions, and eye-catching animations. Flex also provides powerful data handling for industrial-strength applications. We think it should be just as much fun to learn Flex as it is to use it. And we know that fun learning gets better results. Hello! Flex 4 demonstrates how to get started without getting bogged down in technical detail or academic edge cases. In this book, User Friendly cartoon characters offer commentary and snide side comments, as the book moves quickly from Hello World into practical techniques. Each one is illustrated with a hands-on example. Along the way, readers will build a unique Flex application that mashes Yahoo Maps with Twitter to keep track of friends. Purchase of the print book comes with an offer of a free PDF, ePub, and Kindle eBook from Manning. Also available is all code from the book.
              </p>
              <div>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/next.svg"
                    width="22px"
                  />
                </button>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/trash.svg"
                    width="22px"
                  />
                </button>
              </div>
            </div>
            <div>
              <p>
                8:41 AM
              </p>
              <img
                alt="Hello! Flex 4"
                src="http://books.google.com/books/content?id=DjszEAAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
              />
            </div>
          </a>
          <a
            href="/books/16"
          >
            <div>
              <h2
                class="undefined undefined "
              >
                LIFE23333
              </h2>
              <p>
                LIFE Magazine is the treasured photographic magazine that chronicled the 20th Century. It now lives on at LIFE.com, the largest, most amazing collection of professional photography on the internet. Users can browse, search and view photos of todayâ€™s people and events. They have free access to share, print and post images for personal use.
              </p>
              <div>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/next.svg"
                    width="22px"
                  />
                </button>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/trash.svg"
                    width="22px"
                  />
                </button>
              </div>
            </div>
            <div>
              <p>
                8:42 AM
              </p>
              <img
                alt="LIFE23333"
                src="http://books.google.com/books/content?id=51YEAAAAMBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
              />
            </div>
          </a>
          <a
            href="/books/17"
          >
            <div>
              <h2
                class="undefined undefined "
              >
                LIFE23333
              </h2>
              <p>
                LIFE Magazine is the treasured photographic magazine that chronicled the 20th Century. It now lives on at LIFE.com, the largest, most amazing collection of professional photography on the internet. Users can browse, search and view photos of todayâ€™s people and events. They have free access to share, print and post images for personal use.
              </p>
              <div>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/next.svg"
                    width="22px"
                  />
                </button>
                <button
                  class="undefined undefined
  undefined undefined
  undefined "
                  type="button"
                >
                  <img
                    alt="This is the left icon for the button"
                    height="25px"
                    src="assets/icons/trash.svg"
                    width="22px"
                  />
                </button>
              </div>
            </div>
            <div>
              <p>
                8:42 AM
              </p>
              <img
                alt="LIFE23333"
                src="http://books.google.com/books/content?id=51YEAAAAMBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
              />
            </div>
          </a>
        </div>
        <div>
          <button
            class="undefined undefined
  function small() { [native code] } undefined
  undefined "
            data-testid="pagination"
            type="button"
          >
            1
          </button>
          <button
            class="undefined undefined
  function small() { [native code] } undefined
  undefined "
            data-testid="pagination"
            type="button"
          >
            2
          </button>
          <button
            class="undefined undefined
  function small() { [native code] } undefined
  undefined "
            data-testid="pagination"
            type="button"
          >
            3
          </button>
          <button
            class="undefined undefined
  function small() { [native code] } undefined
  undefined "
            data-testid="pagination"
            type="button"
          >
            4
          </button>
        </div>
      </div>
    </section>
  </main>
</div>
`;
